feat: Optionally remove seconds from datetime

gbm001 opened the pull request:
(reopen of #7244 which I couldn't reopen for some reason, probably because it didn't see the force push while closed)

Features: Allows a choice of time format, as well as a choice of date format. Currently HH:mm:ss and HH:mm available (to remove seconds).
Fixes: Updates datepicker to fix bug in datepicker for times with seconds between 0 and 9 (taken from updated Frappe-controlled fork)

Updates geo information to add time_format where date_format already exists (e.g. Country).
Updates server-side formatters and client-side formatters (and datepicker when not using seconds).
Slightly changes datepicker positioning logic.

Includes server-side and Cypress tests.

Description from original PR:
This pull request implements two options for time format in System Settings (very similar to the current date format option). Either HH:mm:ss (equivalent to the current situation) or HH:mm (no seconds) can be selected.
The company I work for has recently started using the Timesheets but there a number of usability issues, namely that using the datetime pickers are very inconvenient to use (particularly the sliders). One issue that has been picked up is the necessity to write ':00' every time you write a time for the seconds (when does a business really ever need to enter seconds?).
Currently frappe datetime and time fields include seconds and have sliders on the time pickers. This pull removes seconds from the pickers if the time format is HH:mm. It also adjusts various client-side formatters to ensure that HH:mm is displayed. The DB continues to use HH:mm:ss in the background.
When using HH:mm:ss, there should only be two changes to current behaviour:

The date pickers now open upwards, rather than downwards, if there is space.
In formatter.js, Datetime had a hard-coded time format of 'h:mm a z' which matched nothing else; it now uses 24hr time but doesn't print the timezone stamp. This is so the displayed time when using a Datetime in a table has consistent formatting between the input itself (clicked on that row, uses the formatting in datetime.js) and the static area (not editing that row; uses formatters.js).

Documentation PR: https://github.com/frappe/erpnext_com/pull/615

gbm001 committed:
fix: Add updated datepicker; fixed seconds formatting bug.
Seconds between 0 and 9 were not zero-padded.

gbm001 committed:
feat: Add framework for time format

gbm001 committed:
feat: datetime server-side formatters.

gbm001 committed:
tests: Added server-side datetime formatter tests

gbm001 committed:
feat: Update client-side datetime formatters

gbm001 committed:
tests: Add Cypress client-side formatting tests.

gbm001 committed:
fix: JSON errors

gbm001 committed:
fix: Update to not hard-code admin password

gbm001 committed:
fix: Change to using bulk_update rather than the REST API

gbm001 committed:
tests: Use Custom doctype for testing, not Standard

gbm001 committed:
fix: Codacy style fixes

gbm001 commented:
The major issue is that despite my best efforts the Cypress tests still take around 6 minutes... This could be reduced by reducing the test coverage, but it can't really be sped up a lot (the datepicker is slow to appear, and wipes out entered text when it does, so it is necessary to give it time to appear before entering text).
Also it was failing on coverage at one point which is clearly rubbish given there is a lot more test code, testing new and old things, than there is non-test code!

gbm001 commented:
Bump...

It's not too surprising that most of the ERPNext code doesn't come from the community given how hard it is to get stuff in...

netchampfaris commented on frappe/core/doctype/system_settings/system_settings.json:
```diff
@@ -117,6 +117,14 @@
   },
+  {
+   "default": "HH:mm:ss",
+   "fieldname": "time_format",
```
This field is shown in the last section. It should go below **Date Format** field.

![image](https://user-images.githubusercontent.com/9355208/68081815-c13b7100-fe39-11e9-8fc2-ca76b8a6afc3.png)

gbm001 replied:
Will be fixed (due to new improved JSON format)

netchampfaris commented on frappe/public/js/frappe/form/controls/date.js:
```diff
@@ -75,12 +75,13 @@ frappe.ui.form.ControlDate = frappe.ui.form.ControlData.extend({
+		var position = 'bottom left';
+		//if (picker_top > scroll_limit + window_scroll) {
+		if (picker_bottom >= scroll_limit) {
+			position = 'top left';
```
What does these changes do? Before / After screenshots?

gbm001 replied:
The old system preferred to open 'down' when possible, which tends to cover the next box (since people work from top to bottom). It also relied on a fixed magic number of 270.
![OLD_prefer_down](https://user-images.githubusercontent.com/4930097/68622034-137b3280-04c9-11ea-8092-3aee1a1fbf1f.png)
It also tended to get it wrong and overlap at the bottom of the window when it should have opened upwards.
In each image below the bottom of the image is the bottom of the window.
Overlapping date picker:
![OLD_overlapping_date](https://user-images.githubusercontent.com/4930097/68621936-d57e0e80-04c8-11ea-9bf0-2e155ff1b2b9.png)
Overlapping time picker:
![OLD_overlapping_time](https://user-images.githubusercontent.com/4930097/68622044-1aa24080-04c9-11ea-8907-0b25a98f323f.png)
Overlapping datetime picker:
![OLD_overlapping_datetime](https://user-images.githubusercontent.com/4930097/68622048-1e35c780-04c9-11ea-9226-08a9cc7bf3b9.png)

gbm001 replied:
New method is (almost) entirely positioned according to screen elements (except fix for broken time-picker which is off by 42px when opening up) and correctly opens down instead of up when there is not space below the header.
Not quite enough space to open 'up' (so opens down, which overlaps bottom of the window slightly as the picker is too long to fit either way):
![NEW_go_down_just](https://user-images.githubusercontent.com/4930097/68622261-a74cfe80-04c9-11ea-8e07-3e8f8747351a.png)
Slightly scrolled up window with just enough space to open upwards:
![NEW_go_up_just](https://user-images.githubusercontent.com/4930097/68622288-b6cc4780-04c9-11ea-9bb1-4d6ee7d65764.png)

netchampfaris commented on frappe/public/js/frappe/form/controls/datetime.js:
```diff
@@ -22,5 +25,37 @@ frappe.ui.form.ControlDatetime = frappe.ui.form.ControlDate.extend({
+		const picker_bottom = this.$input.offset().top + this.$input.outerHeight() + 12 + this.datepicker.$datepicker.outerHeight();
+
+		var position = 'bottom left';
+		//if (picker_top > scroll_limit + window_scroll) {
```
This looks similar to the one in `date.js`. Can we commonify it?

gbm001 replied:
`update_datepicker` can be put in date.js and extended from there to datetime.js and time.js.
`set_datepicker` is not worth it (datetime already calls 'super', the version in time.js is slightly different).

netchampfaris commented on frappe/public/js/frappe/form/controls/datetime.js:
```diff
@@ -22,5 +25,37 @@ frappe.ui.form.ControlDatetime = frappe.ui.form.ControlDate.extend({
+			// it should be due to hiding seconds
+			if (this.noSeconds) this.datepicker.opts['offset'] = -28;
+		} else {
+			if (this.noSeconds) this.datepicker.opts['offset'] = 12;
```
Do these values work properly in all screen sizes? Mobile?

gbm001 replied:
This will need a bit more checking (it's been a while since I last tried it and it was a different ERPNext version).

gbm001 replied:
This has been rejigged a bit. There is now only one 'magic' number to work around mispositioning of the timepicker when opening upwards (it was out by 42px or so). All other numbers (such as the position of the bottom of the header) are taken from the elements. I have tested this in the Chrome devtools 'mobile' simulator thing and it looks like it all works and there is no significant difference.

netchampfaris commented on frappe/public/js/frappe/form/controls/time.js:
```diff
@@ -1,42 +1,87 @@
+			$tp.$secondsText.css('display', 'none');
+			$tp.$secondsText.prev().css('display', 'none');
+		} else {
+			this.noSeconds = false;
```
```suggestion
			this.no_seconds = false;
```

gbm001 replied:
Will be done.

gbm001 replied:
This is now removed anyway.

netchampfaris commented on frappe/public/js/frappe/form/controls/time.js:
```diff
@@ -1,42 +1,87 @@
+			// it should be due to hiding seconds
+			if (this.noSeconds) this.datepicker.opts['offset'] = -28;
+		} else {
+			if (this.noSeconds) this.datepicker.opts['offset'] = 12;
```
I think `time.js` can be extended from `date.js` so we can share the `update_datepicker_position` code.

gbm001 replied:
New plan is to only put update_datepicker_position in date.js, and extend from there to time.js and datetime.js, as suggested.

netchampfaris commented on frappe/public/js/frappe/utils/datetime.js:
```diff
@@ -98,23 +98,28 @@ $.extend(frappe.datetime, {
 		return moment().endOf("year").format();
 	},
 
-	get_user_fmt: function() {
```
The new methods `get_user_date_fmt` and `get_user_time_fmt` are better but we can't lose this old method as it will be a breaking change.

gbm001 replied:
Previous method `get_user_fmt` will be restored (with a comment indicating it is for backwards compatibility only).

netchampfaris commented on frappe/public/js/lib/datepicker/datepicker.min.js:
```diff
@@ -1,2 +1,2 @@
-!function(t,e,s){!function(){var i,a,n,h="2.2.3",o="datepicker",r=".datepicker-here",c=!1,d='<div class="datepicker"><i class="datepicker--pointer"></i><nav class="datepicker--nav"></nav><div class="datepicker--content"></div></div>',l={classes:"",inline:!1,language:"ru",startDate:new Date,firstDay:"",weekends:[6,0],dateFormat:"",altField:"",altFieldDateFormat:"@",toggleSelected:!0,keyboardNav:!0 ...
-this.$ranges.on("mousemove focus ",this._onMouseEnterRange.bind(this)),this.$ranges.on("mouseout blur",this._onMouseOutRange.bind(this))},_setTime:function(t){var e=i.getParsedDate(t);this._handleDate(t),this.hours=e.hours<this.minHours?this.minHours:e.hours,this.minutes=e.minutes<this.minMinutes?this.minMinutes:e.minutes,this.seconds=e.seconds<this.minSeconds?this.minSeconds:e.seconds},_setMinTi ...
\ No newline at end of file
+!function(t,e,s){!function(){var i,a,n,h="2.2.3",o="datepicker",r=".datepicker-here",c=!1,d='<div class="datepicker"><i class="datepicker--pointer"></i><nav class="datepicker--nav"></nav><div class="datepicker--content"></div></div>',l={classes:"",inline:!1,language:"ru",startDate:new Date,firstDay:"",weekends:[6,0],dateFormat:"",altField:"",altFieldDateFormat:"@",toggleSelected:!0,keyboardNav:!0 ...
```
Where is this compiled from? I'd prefer we make changes on the official fork https://github.com/frappe/air-datepicker and compile it from there.

gbm001 replied:
It was supposed to be from the frappe/air-datepicker repository. Pull request 1 in March (which fixes a datetime display issue for seconds) on that repository never (I think) got merged back into Frappe. This might be better in a separate Pull Request? It just needs the datepicker.min.js file copying from the air-datepicker repository into the Frappe repository.

netchampfaris commented on frappe/utils/data.py:
```diff
@@ -208,15 +208,21 @@ def get_datetime_str(datetime_obj):
 		datetime_obj = get_datetime(datetime_obj)
 	return datetime_obj.strftime(DATETIME_FORMAT)
 
-def get_user_format():
```
Let's not remove this old method for backwards compatibility.

gbm001 replied:
`get_user_format` reinstated (with comment warning) as alias for get_user_date_format.

netchampfaris commented on frappe/website/render.py:
```diff
@@ -125,12 +125,12 @@ def build_response(path, data, http_status_code, headers=None):
 	response.data = set_content_type(response, data, path)
 	response.status_code = http_status_code
-	response.headers["X-Page-Name"] = path.encode("utf-8")
+	response.headers["X-Page-Name"] = path.encode("ascii", errors="xmlcharrefreplace")
```
These changes are already merged. Please rebase.

gbm001 replied:
This has been added by mistake and I will remove it in a rebase as suggested...

netchampfaris commented on cypress/integration/datetime.js:
```diff
@@ -0,0 +1,192 @@
+/* Test date, time and datetime formats.
```
I appreciate that this is an extensive test suite for all things datetime, but it increases the test time by almost 3-4 minutes. Let's reduce this to a few nominal tests. On the UI side we should only test the user input and ensure that the values are saved properly in the form. You already have covered a lot of formats in Python unittest.

gbm001 replied:
Reduced to testing one date format and two time formats.

gbm001 replied:
Reduced to testing a single date format and both time formats.

netchampfaris commented:
@gbm001 I apologize that I took this long to review this. Let's get this merged 💪🏼.

gbm001 committed:
fix: Commonify update_datetime_picker in date.js, datetime.js, time.js
Fix order of time_format in System Settings
Restore get_user_fmt in utils/datetime.js

gbm001 committed:
feat: Drastically reduce scale of Cypress testing (to make tests faster)
Full testing is possible by setting 'fast_mode' to false in the spec file.

gbm001 committed:
fix: Fix issues with datepicker/timepicker expansion

gbm001 committed:
fix: typo

gbm001 committed:
style: Various style fixes as requested by DeppSource: Python

gbm001 committed:
fix: Timepicker not hiding on 'now' button. Force hiding on click.

gbm001 committed:
style: Codacy style fixes.

netchampfaris committed:
fix: Use datepicker from node_modules

gbm001 commented:
So what happens now? I'm not sure what to do with the yarn.lock conflict; I assume it wants to mirror current develop but I'm not even sure what Yarn does so probably shouldn't start guessing :)

gbm001 commented:
bump...

surajshetty3416 committed:
Merge branch 'develop' into gbm001-time-format2

netchampfaris committed:
test: Refactor Datetime UI tests

- cy.get_field
- cy.set_value
- cy.insert_doc with ignore_duplicate
- Nominal datetime tests to cover most formats
- Formatting with prettier

netchampfaris commented:
@gbm001 This is almost ready to be merged. Can you update the docs [here](https://github.com/frappe/erpnext_com/blob/master/erpnext_com/www/docs/user/manual/en/setting-up/settings/system-settings.md) for the Time Format field?

Edit: Opened a PR: https://github.com/frappe/erpnext_com/pull/615

gbm001 commented:
Thanks for doing the docs :)

gbm001 committed:
test: Datetime UI tests; wait for cur_frm.doc.datetime to update

gbm001 commented:
My last commit changed my local Cypress runs from reliably failing at the same point as Travis to reliably passing, but apparently not a good enough fix to stop it being flaky in Travis...

netchampfaris commented:
@gbm001 Yeah, it's weird because `datetime.js` and `table_multiselect.js` are consistently failing in Travis.

netchampfaris committed:
tests: Add whitespace to typed input

- Clear input only for Time field

netchampfaris committed:
test: Wait timeout 200

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/frappe into gbm001-time-format2

netchampfaris committed:
test: Fix form test

surajshetty3416 committed:
Merge branch 'develop' into gbm001-time-format2

surajshetty3416 committed:
Merge branch 'develop' into gbm001-time-format2

surajshetty3416 committed:
Merge branch 'develop' into gbm001-time-format2

surajshetty3416 committed:
Merge branch 'develop' into gbm001-time-format2

netchampfaris commented:
@gbm001 🎉

gbm001 commented:
Took a while to make it work, but got there in the end :)
